lib/newspaper_works/ingest/pub_finder.rb

```diff
@@ -0,0 +1,82 @@
+require 'newspaper_works/logging'
+require 'newspaper_works/ingest'
+
+module NewspaperWorks
+  module Ingest
+    # mixin for find-or-create of publication, for use by various ingests
+    module PubFinder
+      include NewspaperWorks::Logging
+
+      COPY_FIELDS = [
+        :title,
+        :lccn,
+        :oclcnum,
+        :issn,
+        :place_of_publication,
+        :language,
+        :preceded_by,
+        :succeeded_by
+      ].freeze
+
+      MULTI_VALUED = [
+        :title,
+        :language,
+        :preceded_by,
+        :succeeded_by,
+        :place_of_publication
+      ].freeze
+
+      # @param lccn [String] Library of Congress Control Number
+      #   of Publication
+      # @return [NewspaperTitle, NilClass] publication or nil if not found
+      def find_publication(lccn)
+        NewspaperTitle.where(lccn: lccn).first
+      end
+
+      # Copy publication metadata from authority lookup for LCCN
+      # @param publication [NewspaperTitle]
+      # @param metadata [NewspaperWorks::Ingest::PublicationInfo]
+      def copy_publication_metadata(publication, metadata, lccn, title = nil)
+        COPY_FIELDS.each do |name|
+          value = metadata.send(name)
+          next if value.nil?
+          value = [value] if MULTI_VALUED.include?(name)
+          publication.send("#{name}=", value)
+        end
+        # prefer locally-specified title to looked-up title:
+        publication.title = [title] unless title.nil?
+        # final fallback, nothing specified, title mandatory: use LCCN
+        publication.title = [lccn] if publication.title.empty?
+      end
+
+      def create_publication(lccn, title = nil, opts = {})
+        publication = NewspaperTitle.create
+        info = NewspaperWorks::Ingest::PublicationInfo.new(lccn)
+        copy_publication_metadata(publication, info, lccn, title)
+        publication.lccn ||= lccn
+        NewspaperWorks::Ingest.assign_administrative_metadata(publication, opts)
+        publication.save!
+        write_log(
+          "Created NewspaperTitle work #{publication.id} for LCCN #{lccn}"
+        )
+        publication
+      end
+
+      def find_or_create_publication_for_issue(issue, lccn, title, opts)
+        publication = find_publication(lccn)
+        unless publication.nil?
+          write_log(
+            "Found existing NewspaperTitle #{publication.id}, LCCN #{lccn}"
+          )
+        end
+        publication = create_publication(lccn, title, opts) if publication.nil?
+        publication.ordered_members << issue
```

Comment:
Do we want to use `ordered_members` here? Have we used this in the past for NDNP? My recollection is that issues are related to publications by `members` but not `ordered_members`.

Reply:
This is a good question.  This is a holdover from how it was done with NDNP ingest, but there is  not a reason I can think of to need the ordering.

I suspect that using members is reasonable — the only frustration I have with these associations is that `ordered_members` and `members` are wholly disjoint.  Ideally, there would be an accessor that allowed access to the combined set (vs. cognitive load of having to remembers to check each, both, or one or the other).

Reply:
This and other feedback should be resolved by commit 208438424f64de1ac9900c6be36b1a918f29494f